File: Python/problem0414.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jun  6 12:04:26 2019

@author: Administrator
"""

class Solution:
    def thirdMax(self, nums: list) -> int:
        nums = list(set(nums))
        if not nums:
            return None
        if len(nums)<3:
            return max(nums)
        for m in range(3):
            Max = nums[0]
            for k in nums:
                Max = max(Max, k)
            if m != 3-1:
                nums.remove(Max)
            else:
                return Max
        
        
solu = Solution()
nums = [3, 2, 1]
#nums = [1, 2]
nums = [2,2,3,1]
print(solu.thirdMax(nums))

File: C++/problem0414.cpp
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        // sort(nums.begin(), nums.end());

        set<int> s;
        for (vector<int>::iterator it = nums.begin(); it < nums.end(); ++it)
        {
            s.insert(*it);
        }
        
        if (s.size() < 3)
        {
            int max_ = nums[0];
            for (vector<int>::iterator it = nums.begin(); it < nums.end(); ++it)
            {
                if (*it > max_)
                {
                    max_ = *it;
                }
            }
            return max_;
        }

        // vector<int>::iterator it = nums.end();
        // return *(it - 3);

        int First = INT_MIN, Second = INT_MIN, Third = INT_MIN;
        for (vector<int>::iterator it = nums.begin(); it < nums.end(); ++it)
        {
            if (*it == First || *it == Second || *it == Third)
            {
                continue;
            }
            if (*it > First)
            {
                Third = Second;
                Second = First;
                First = *it;
            }
            else if (*it > Second)
            {
                Third = Second;
                Second = *it;
            }
            else if (*it > Third)
            {
                Third = *it;
            }
        }
        return Third;
    }
};
